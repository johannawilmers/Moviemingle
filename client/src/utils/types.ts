/**
 * Interface representing a movie.
 * @interface
 * @property {number | undefined} _id - The unique identifier of the movie.
 * @property {string} title - The title of the movie.
 * @property {string} original_title - The original title of the movie.
 * @property {string} poster_path - The path to the movie's poster image.
 * @property {string} backdrop_path - The path to the movie's backdrop image.
 * @property {number[]} genre_ids - An array of genre IDs associated with the movie.
 * @property {string} overview - A brief overview or description of the movie.
 * @property {string} release_date - The release date of the movie.
 * @property {number} vote_average - The average vote or rating for the movie.
 * @property {boolean | undefined} isFavorite - Whether the movie is marked as a favorite.
 */
export interface movie {
  _id: number | undefined;
  title: string;
  original_title: string;
  poster_path: string;
  backdrop_path: string;
  genre_ids: number[];
  overview: string;
  release_date: string;
  vote_average: number;
  isFavorite?: boolean;
}

/**
 * Interface representing a genre.
 * @interface
 * @property {number | undefined} _id - The unique identifier of the genre.
 * @property {string} name - The name of the genre.
 */
export interface genre {
  _id: number | undefined;
  name: string;
}
