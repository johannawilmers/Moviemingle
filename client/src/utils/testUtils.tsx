import { ApolloClient, ApolloProvider, HttpLink, InMemoryCache } from '@apollo/client';
import { MockedProvider } from '@apollo/client/testing';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

/**
 * HTTP link for Apollo Client to connect to a GraphQL server.
 */
const link = new HttpLink({
  uri: 'http://localhost:4000',
});

/**
 * Apollo Client instance configured with an in-memory cache and HTTP link.
 * @type {import('@apollo/client').ApolloClient<import('@apollo/client').NormalizedCacheObject>}
 */
const client = new ApolloClient({
  cache: new InMemoryCache(),
  link: link,
});

/**
 * Renders a component with Apollo Client and BrowserRouter for testing page content.
 * @function
 * @param {JSX.Element} component - React component to render.
 * @returns {RenderResult} - Result of rendering with testing utilities.
 */
export const renderWithPageContent = (component: JSX.Element) =>
  render(
    <ApolloProvider client={client}>
      <BrowserRouter>{component}</BrowserRouter>
    </ApolloProvider>,
  );

/**
 * Renders a component with Apollo MockedProvider for testing with mocked GraphQL data.
 * @function
 * @param {React.ReactNode} child - React component to render within the provider.
 * @param {any[]} mocks - Array of GraphQL query/mutation mocks.
 * @returns {RenderResult} - Result of rendering with testing utilities.
 */
export const renderWithMocks = (child: React.ReactNode, mocks: any[] = []) => {
  return render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <BrowserRouter>{child}</BrowserRouter>
    </MockedProvider>,
  );
};
