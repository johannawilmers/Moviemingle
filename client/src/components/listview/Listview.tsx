import './listview.css';

import { movie } from '../../utils/types';
import { MovieCard } from '../movieCard/movieCard';
interface props {
  movies: movie[];
}
export const ListView = ({ movies }: props) => {
  const userID = localStorage.getItem('userID');

  if (movies.length === 0) {
    return <p id="errorMessageSearch"> There are no movies matching this search </p>;
  }

  return (
    <div className="listview-container">
      {movies.map((movie: movie) => (
        <MovieCard key={movie._id} movie={movie} userID={userID || ''} />
      ))}
    </div>
  );
};
export default ListView;
