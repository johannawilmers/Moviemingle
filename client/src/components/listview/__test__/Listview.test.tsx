import { screen } from '@testing-library/react';

import { renderWithPageContent } from '../../../utils/testUtils';
import { movie } from '../../../utils/types';
import { ListView } from '../Listview';

describe('ListView', () => {
  const movies: movie[] = [
    {
      backdrop_path: '/4XM8DUTQb3lhLemJC51Jx4a2EuA.jpg',
      genre_ids: [28, 80, 53],
      _id: 385687,
      original_title: 'Fast X',
      overview:
        "Over many missions and against impossible odds, Dom Toretto and his family have outsmarted, out-nerved and outdriven every foe in their path. Now, they confront the most lethal opponent they've ever faced: A terrifying threat emerging from the shadows of the past who's fueled by blood revenge, and who is determined to shatter this family and destroy everything—and everyone—that Dom loves, forever.",
      poster_path: '/fiVW06jE7z9YnO4trhaMEdclSiC.jpg',
      release_date: '2023-05-17',
      title: 'Fast X',
      vote_average: 7.3,
      isFavorite: true,
    },
    {
      backdrop_path: '/iIvQnZyzgx9TkbrOgcXx0p7aLiq.jpg',
      genre_ids: [27, 53],
      _id: 1008042,
      original_title: 'Talk to Me',
      overview:
        'When a group of friends discover how to conjure spirits using an embalmed hand, they become hooked on the new thrill, until one of them goes too far and unleashes terrifying supernatural forces.',
      poster_path: '/kdPMUMJzyYAc4roD52qavX0nLIC.jpg',
      release_date: '2023-07-26',
      title: 'Talk to Me',
      vote_average: 7.2,
      isFavorite: false,
    },
  ];

  it('should match snapshot', () => {
    const { container } = renderWithPageContent(<ListView movies={movies} />);
    expect(container).toMatchSnapshot();
  });

  it('should render one picture for each movie in the list', () => {
    renderWithPageContent(<ListView movies={movies} />);
    const movieCards = screen.getAllByRole('img');
    expect(movieCards).toHaveLength(2);
  });
});
