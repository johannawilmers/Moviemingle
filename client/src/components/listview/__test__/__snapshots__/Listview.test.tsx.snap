// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ListView > should match snapshot 1`] = `
<div>
  <div
    class="listview-container"
  >
    <div
      class="movie-container"
    >
      <a
        class="link"
        href="/movie/385687"
        id="movieImage-385687"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root movie-card css-bhp9pd-MuiPaper-root-MuiCard-root"
        >
          <img
            alt="Fast X"
            id="movieCardImage"
            src="https://image.tmdb.org/t/p/w200//fiVW06jE7z9YnO4trhaMEdclSiC.jpg"
          />
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-d3wcwz-MuiTypography-root"
            >
              <p>
                Fast X
              </p>
            </div>
            <div
              class="rating"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="StarIcon"
                focusable="false"
                style="font-size: 32px;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
              <p>
                7.3
                 / 10
              </p>
            </div>
          </div>
        </div>
      </a>
      <div
        class="heartSection"
        id="favButton-385687"
      >
        <div>
          <button
            aria-label="Remove movie from favorites"
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            id="favButton"
            style="cursor: pointer; z-index: 10; position: relative;"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="FavoriteIcon"
              focusable="false"
              style="font-size: 44px;"
              viewBox="0 0 24 24"
            >
              <path
                d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      class="movie-container"
    >
      <a
        class="link"
        href="/movie/1008042"
        id="movieImage-1008042"
      >
        <div
          class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root movie-card css-bhp9pd-MuiPaper-root-MuiCard-root"
        >
          <img
            alt="Talk to Me"
            id="movieCardImage"
            src="https://image.tmdb.org/t/p/w200//kdPMUMJzyYAc4roD52qavX0nLIC.jpg"
          />
          <div
            class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <div
              class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-d3wcwz-MuiTypography-root"
            >
              <p>
                Talk to Me
              </p>
            </div>
            <div
              class="rating"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="StarIcon"
                focusable="false"
                style="font-size: 32px;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 17.27 18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
              </svg>
              <p>
                7.2
                 / 10
              </p>
            </div>
          </div>
        </div>
      </a>
      <div
        class="heartSection"
        id="favButton-1008042"
      >
        <div>
          <button
            aria-label="Add movie to favorites"
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            id="favButton"
            style="cursor: pointer; z-index: 10; position: relative;"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="FavoriteBorderIcon"
              focusable="false"
              style="font-size: 44px;"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
